### PR TITLE
Update gsgnn_node_emb for multi-task learning

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#789 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 1]
```diff
@@ -63,36 +70,51 @@ def main(config_args):
...
+    if config.multi_tasks:
+        # Only support multi-task sahred encoder model.
+        model = GSgnnMultiTaskSharedEncoderModel(config.alpha_l2norm)
+        gs.gsf.set_encoder(model, input_data.g, config, train_task=False)
+        assert config.restore_model_layers is not GRAPHSTORM_MODEL_ALL_LAYERS, \
+            "When computing node embeddings with GSgnnMultiTaskSharedEncoderModel, " \
+            "please set --restore-model-layers to " \
+            f"{GRAPHSTORM_MODEL_EMBED_LAYER}, {GRAPHSTORM_MODEL_GNN_LAYER}." \
+            f"Please do not include {GRAPHSTORM_MODEL_DECODER_LAYER}" \
+            f"But we get {config.restore_model_layers}"
```
Comment (jalencato) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:82]:
```suggestion
            f"but we get {config.restore_model_layers}"
```

File: tests/end2end-tests/graphstorm-mt/mgpu_test.sh [hunk 2]
```diff
@@ -190,3 +190,18 @@ then
...
+
+cnt=$(ls -l /data/gsgnn_mt/save-emb/ | wc -l)
+cnt=$[cnt - 1]
+if test $cnt != 2
+then
+    echo "The number of saved embs $cnt is not equal to 2 (for movie and user)."
+fi
+
+# Multi-task will save node embeddings of all the nodes.
+python3 $GS_HOME/tests/end2end-tests/check_infer.py --train-embout /data/gsgnn_mt/emb/ --infer-embout /data/gsgnn_mt/save-emb/ --link-prediction
```
Comment (jalencato) [inline tests/end2end-tests/graphstorm-mt/mgpu_test.sh:207]: Do we need to check if the stored infer_ntypes are sorted?

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 3]
```diff
@@ -63,36 +70,51 @@ def main(config_args):
...
-            infer_ntypes.add(etype[2])
-        infer_ntypes = sorted(list(infer_ntypes))
     else:
-        raise TypeError("Not supported for task type: ", task_type)
+        task_type = config.task_type
+        # infer ntypes must be sorted for node embedding saving
+        if task_type == BUILTIN_TASK_LINK_PREDICTION:
+            infer_ntypes = None
+        elif task_type in {BUILTIN_TASK_NODE_REGRESSION, BUILTIN_TASK_NODE_CLASSIFICATION}:
+            # TODO(xiangsx): Support multi-task on multiple node types.
```
Comment (jalencato) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:108]: May we remove the comment here?

**Review by jalencato — COMMENTED** [review]
Some small comment.

File: tests/end2end-tests/graphstorm-mt/mgpu_test.sh (hunk 2 above)
Comment (classicsong) [reply tests/end2end-tests/graphstorm-mt/mgpu_test.sh:207]: No. As sorting is done by the remap script.
Here we only test if the embeddings saved in the training process is same as the embeddings saved in gen_embedding are same.

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py (hunk 3 above)
Comment (classicsong) [reply python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:108]: That makes sense.

**Review by jalencato — APPROVED** [review]
LGTM

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 4]
```diff
@@ -63,36 +70,50 @@ def main(config_args):
...
 
     # load the model
-    if config.task_type == BUILTIN_TASK_LINK_PREDICTION:
-        model = gs.create_builtin_lp_gnn_model(input_data.g, config, train_task=False)
-    elif config.task_type in {BUILTIN_TASK_NODE_REGRESSION, BUILTIN_TASK_NODE_CLASSIFICATION}:
-        model = gs.create_builtin_node_gnn_model(input_data.g, config, train_task=False)
-    elif config.task_type in {BUILTIN_TASK_EDGE_CLASSIFICATION, BUILTIN_TASK_EDGE_REGRESSION}:
-        model = gs.create_builtin_edge_gnn_model(input_data.g, config, train_task=False)
+    if config.multi_tasks:
+        # Only support multi-task sahred encoder model.
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:74]: sahred -> shared

File: python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py [hunk 5]
```diff
@@ -63,36 +70,50 @@ def main(config_args):
...
+        model = GSgnnMultiTaskSharedEncoderModel(config.alpha_l2norm)
+        gs.gsf.set_encoder(model, input_data.g, config, train_task=False)
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:76]: can we use `gs.set_encoder` here?

Comment (classicsong) [reply python/graphstorm/run/gsgnn_emb/gsgnn_node_emb.py:76]: set_encoder is not included in gsf.__init__.

**Pull request merged**